### PR TITLE
8306331: assert((cnt > 0.0f) && (prob > 0.0f)) failed: Bad frequency assignment in if

### PR DESCRIPTION
**Pull request by dean-long**
This change removes undefined behavior caused by signed overflow, which triggered an assert with Xcode14.3+1.0-beta1 on macos aarch64.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306331](https://bugs.openjdk.org/browse/JDK-8306331): assert((cnt &gt; 0.0f) &amp;&amp; (prob &gt; 0.0f)) failed: Bad frequency assignment in if

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [9e7b087b](https://git.openjdk.org/jdk/pull/13551/files/9e7b087b1b02234181f3f3397d51a0bc552a1263)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13551/head:pull/13551` \
`$ git checkout pull/13551`

Update a local copy of the PR: \
`$ git checkout pull/13551` \
`$ git pull https://git.openjdk.org/jdk.git pull/13551/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13551`

View PR using the GUI difftool: \
`$ git pr show -t 13551`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13551.diff">https://git.openjdk.org/jdk/pull/13551.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13551#issuecomment-1515636973)

### CODE REVIEW

File: src/hotspot/share/opto/parse2.cpp [hunk 1]
```diff
@@ -1192,6 +1192,25 @@ static bool has_injected_profile(BoolTest::mask btest, Node* test, int& taken, i
   }
   return false;
 }
+
+// Give up if too few (or too many, in which case the sum will overflow) counts to be meaningful.
+// We also check that individual counters are positive first, otherwise the sum can become positive.
+// (check for saturation, integer overflow, and immature counts)
+static bool counters_are_meaningful(int counter1, int counter2, int min) {
+  // check for saturation, inluding "uint" values too big to fit it "int"
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/parse2.cpp:1200]:
```suggestion
  // check for saturation, including "uint" values too big to fit in "int"
```

File: src/hotspot/share/opto/parse2.cpp [hunk 2]
```diff
@@ -1192,6 +1192,25 @@ static bool has_injected_profile(BoolTest::mask btest, Node* test, int& taken, i
...
+    return false;
+  }
+  // check for integer overflow of the sum
+  int64_t sum = (int64_t)counter1 + (int64_t)counter2;
+  STATIC_ASSERT(sizeof(counter1) < sizeof(sum));
+  if (sum > INT_MAX) {
+    return false;
+  }
+  // check if mature
+  return counter1 + counter2 >= min;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/parse2.cpp:1211]:
```suggestion
  return (counter1 + counter2) >= min;
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by dean-long** [comment]
Thanks Tobias.

**Comment by dean-long** [comment]
After looking at the code again, I was wondering if anyone prefers using "count" instead of "counter" here, so this:

`static bool counters_are_meaningful(int counter1, int counter2, int min) {`

would become this:

`static bool counts_are_meaningful(int count1, int count2, int min) {`

**Comment by TobiHartmann** [comment]
Sounds reasonable to me.

**Comment by dean-long** [comment]
I think I'll just leave the names as-is.  The comments in the code seem to use "count" and "counter" interchangeably.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Comment by dean-long** [comment]
Thanks Christian.

**Pull request closed without merging**